test/core/storage/keystore/filesystem_keystore_test.cpp

```diff
@@ -0,0 +1,318 @@
+}
+
+/**
+ * @given Keystore is empty
+ * @when call list
+ * @then empty list returned
+ */
+TEST_F(FileSystemKeyStoreTest, ListEmpty) {
+  EXPECT_OUTCOME_TRUE(list, ks->List());
+  ASSERT_EQ(0, list.size());
+}
+
+/**
+ * @given Keystore stores 2 keys
+ * @when call list
+ * @then list containing all addresses returned
+ */
+TEST_F(FileSystemKeyStoreTest, ListKeys) {
+  EXPECT_OUTCOME_TRUE_1(ks->Put(bls_address_, bls_keypair_->private_key));
+  EXPECT_OUTCOME_TRUE_1(
+      ks->Put(secp256k1_address_, secp256k1_keypair_->private_key));
+
+  EXPECT_OUTCOME_TRUE(list, ks->List());
+  ASSERT_EQ(2, list.size());
+  ASSERT_TRUE(VectorContains(list, bls_address_));
+  ASSERT_TRUE(VectorContains(list, secp256k1_address_));
+}
+
+/**
+ * @given empty Keystore
+ * @when sign with wrong address
+ * @then NOT_FOUND returned
+ */
+TEST_F(FileSystemKeyStoreTest, SignNotFound) {
+  auto res = ks->Sign(bls_address_, data_);
+  ASSERT_FALSE(res);
+  ASSERT_EQ(KeyStoreError::NOT_FOUND, res.error());
+}
+
+/**
+ * @given empty Keystore
+ * @when put malformed address
+ * @then WRONG_ADDRESS returned
+ */
+TEST_F(FileSystemKeyStoreTest, SignWrongAddress) {
+  // generate id address (type 0) which has no key
+  std::vector<uint8_t> bytes{0x0, 0x0, 0xD1, 0xC2, 0xA7, 0x0F};
+  auto wrong_address = fc::primitives::decode(bytes).value();
+  auto res = ks->Put(wrong_address, bls_keypair_->private_key);
+  ASSERT_FALSE(res);
+  ASSERT_EQ(KeyStoreError::WRONG_ADDRESS, res.error());
+}
+
+/**
+ * @given Kestore with bls private key
+ * @when Sign() called with bls crypto
+ * @then correct signature returned
+ */
+TEST_F(FileSystemKeyStoreTest, SignCorrectBls) {
+  EXPECT_OUTCOME_TRUE_1(ks->Put(bls_address_, bls_keypair_->private_key));
+  EXPECT_OUTCOME_TRUE(signature, ks->Sign(bls_address_, data_));
+  auto bls_signature = boost::get<BlsSignature>(signature);
+  ASSERT_TRUE(checkSignature(data_, bls_signature, bls_keypair_->public_key));
+}
+
+/**
+ * @given Kestore with secp256k1 private key
+ * @when Sign() called with bls crypto
+ * @then correct signature returned
+ */
+TEST_F(FileSystemKeyStoreTest, SignCorrectSecp256k1) {
+  EXPECT_OUTCOME_TRUE_1(
+      ks->Put(secp256k1_address_, secp256k1_keypair_->private_key));
+  EXPECT_OUTCOME_TRUE(signature, ks->Sign(secp256k1_address_, data_));
+  auto secp256k1_signature = boost::get<Secp256k1Signature>(signature);
+  ASSERT_TRUE(checkSignature(
+      data_, secp256k1_signature, secp256k1_keypair_->public_key));
+}
+
+/**
+ * @given empty Keystore
+ * @when Verify() with wrong address
+ * @then NOT_FOUND returned
+ */
+TEST_F(FileSystemKeyStoreTest, VerifyNotFound) {
+  BlsSignature signature;
+  auto res = ks->Verify(bls_address_, data_, signature);
+  ASSERT_FALSE(res);
+  ASSERT_EQ(KeyStoreError::NOT_FOUND, res.error());
+}
+
+/**
+ * @given Keystore with bls private key
+ * @when Verify() called with wrong signature
+ * @then false returned
+ */
+TEST_F(FileSystemKeyStoreTest, VerifyWrongBls) {
```

Comment:
why `Verify` behavior for secp256k1 and bls is inconsistent?
for invalid signature VerifyWrongBls test expects `outcome::success(false)` and InvalidSecp256k1Signature test expects `outcome::failure(CryptoProviderError::SIGNATURE_VERIFICATION_FAILED)`

Reply:
Because any bls signature is valid. And test `VerifyWrongBls` verifies wrong signature. In case of secp256k1 signature with zeroes is invalid and signature algorithm fails in test `InvalidSecp256k1Signature`. Valid but wrong secp256k1 is tested in `WrongSecp256k1Signature`.